### DISCUSSION TITLE
Rationale behind not chaining up to Root CA in K3s or how to correctly establish trust

### DISCUSSION THREAD

**Discussion by dispyfree** [k3s-io/k3s#7596] [Q&A]
**Environmental Info:**
K3s Version: 

k3s version v1.26.4+k3s1 (8d0255af)
go version go1.19.8

Node(s) CPU architecture, OS, and Version: 
Linux localhost.localdomain 6.2.13-300.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Apr 27 01:33:30 UTC 2023 x86_64 GNU/Linux

Cluster Configuration:
single node 

### Context
I am currently building a provisioning pipeline for an air-gapped K3s and ran into a problem with trust.
Specifically, I am struggling to understand why requests to the K3s API server only contain certificates up to the "Server CA", but not the "Intermediate CA". To understand what I am referring to here, please check out [the illustration of the certificate hierarchy as presented here][1]. 

### Problem Statement:
Among other things, I access the K3s API from outside of the K3s cluster. To this end, the client needs to trust the K3s API HTTPS connection, i.e. it needs to be able to validate the HTTPS certificate it is presented with.
The K3s API only sends the leaf certificate plus the "Server CA" certificate to the client; for lack of "Intermediate CA", the client then cannot tie these certificates to the "Root CA" certificate that it already trusts (that is in fact the only certificate that I put into the store).
I noted specifically that K3s API only returns these two certificates as part of its response, but not all certificates. Note how "~/.kube/config" for k3s (i.e. configuration used by kubectl) specifically includes the _entire_ certificate chain, so this is not a problem there. 

### Question: 
What is the design rationale to not respond with the entire chain s.t. clients can validate all the way up to "Root CA" (which they already trust)? 
Am I opening myself up to any architectural problem if I somehow get K3s API to send all relevant certificates in the response?

Specifically, I wouldn't want to distribute the "Intermediate CA" for my clients to trust; I would rather want to distribute the public key of the "Root CA". That has a couple of advantages:
- I can rotate the "Intermediate CA" without redistributing to clients
- I can chain the Root CAs a bit more and distribute _one_ "Root CA" to clients so that they automatically trust all clusters which use certificates signed by that "Root CA"  

Thank you in advance!  

  [1]: https://docs.k3s.io/cli/certificate

**Comment by brandond** [comment]
The idea is that only the 5 3rd-tier intermediates CAs (ServerCA and its peers) are unique to the Kubernetes cluster itself; the Root and Intermediate CAs are intended to be part of an existing private PKI system, with CA distribution already handled within your organization.

Generally accepted guidance for private PKI is to set up a root that is kept in a secure, offline location, and then use that to sign an intermediate for day-to-day issuance. The certs for both the root and intermediate should be distributed to clients, and the intermediate used to sign the 3rd-tier CAs for your cluster.

**Reply by dispyfree** [reply]
@brandond  Thank you - that confirms my hunch that the intermediates are to be distributed. 

What is the reason that K3s does not include the 2nd tier intermediate in its response? That would enable clients to only go with the Root CA for trust, no need for the intermediates. 

Note how some implementations will stop certificate validation at the first trusted certificate - if you add 1st tier root and 2nd tier intermediate, they will simply stop validation at the 2nd tier intermediate and never go check the root. Prominent proponents of this are Chrome and Firefox, so this is widely spread. 
**This also means that Root CAs are never checked for revocation** (but correct me if I'm wrong on this), as opposed to the situation where the intermediate is not trusted. 

It's present on the server anyway and the server currently already includes the 3rd tier in its response.

**Reply by brandond** [reply]
> This also means that Root CAs are never checked for revocation (but correct me if I'm wrong on this), as opposed to the situation where the intermediate is not trusted.

Kubernetes and golang in general do not support CRL checks, so there's no point in worrying about checking to see if the root CA has been revoked. There are stubs in the x509 code for clients to implement it, but out of the box golang (and Kubernetes by extension) TLS servers and clients will not check any CRL distribution points listed in CA certs to see if the any of the certs in the chain have been revoked. If that's what you're hoping to enable here, it's not going to work.

Kubernetes does not expect that the server's CA will be in a CA bundle that is already trusted by the client. The expectation is that the client's kubeconfig contains the trust anchor, and the server sends enough to create a full chain.

**Reply by brandond** [reply]
> What is the reason that K3s does not include the 2nd tier intermediate in its response? That would enable clients to only go with the Root CA for trust, no need for the intermediates.

This isn't intentional; the library that we're using for the TLS listener (https://github.com/rancher/dynamiclistener/) just only only loads the first entry from the CA file, so that's all that gets sent along in the TLS handshake.

**Reply by dispyfree** [reply]
@brandond 

- just checked: the client's kubeconfig (as generated by `kubectl config view --raw`) contains the _entire_ CA chain, from the Root CA all the way to `server-ca`. I.e. this doesn't really require you to have anything in your trust store, as it just ships all of them.
... probably me being silly to grab the config view output instead of hand-crafting that connection information 

- do you see value in sending along also the intermediate cert? If yes, then I'd probably pick that up.

**Reply by brandond** [reply]
Sure, we'd probably take that PR. The path you're looking at is at:
* https://github.com/k3s-io/k3s/blob/master/pkg/cluster/https.go#L50
* https://github.com/rancher/dynamiclistener/blob/master/factory/ca.go#L81

Note that these functions all work on `*x509.Certificate`, when what you're wanting to do will require `[]*x509.Certificate` to hold the full chain (or at least more of it). Somewhat like the bits at:
* https://github.com/rancher/dynamiclistener/blob/master/cert/pem.go#L152

**Reply by knoppiks** [reply]
I added a [PR on rancher/dynamiclistener](https://github.com/rancher/dynamiclistener/pull/78), compiled k3s with it and was able to see all intermediate certificates in the chain.

(We are running a PKI using hashicorp/vault and wanted to keep only our Root CA distributed to external services/clients when considering a switch to k3s)
